### ISSUE TITLE
Drawing Child Rows on FF with *minified* JS 10x Slower than unminified JS

### ISSUE THREAD

**Issue by Benno007** [DataTables/DataTables#491]
I've just downloaded datatables to use in my local dev. The basic setup is just 6 columns, 30 rows, each with a child row that's appended on draw.dt (I need all these child rows to always display, instead of when clicking on a + to dropdown details).

In Firefox 35 (Chrome 40 is fine, IE10 appears fine), the minified version of your script is 10x slower on a draw with child rows than the unminified version.
Unminified:
![fast profile](https://cloud.githubusercontent.com/assets/313195/6071562/f2ee604c-ade2-11e4-8271-b08d4e8c2089.png)
Minified:
![slow profile](https://cloud.githubusercontent.com/assets/313195/6071563/f2ee6eac-ade2-11e4-9faf-3de4fe946edd.png)

Open these in Firefox with Firebug open and you should see the profile run
JsFiddle with Minified JS: http://jsfiddle.net/w21gzbxe/1/
JsFiddle with Unminified JS: http://jsfiddle.net/ddpL3cw3/

The offending function appears to be _Api.extend.

I think it could be to do with either the way the minified version has its condition remapped, or how the for loop works and removes the braces. 

My solution was to copy the unminified version, remap the variable names to their minified counterparts, and replace the function in the minified file, and it made it faster. But maybe there's something you can do in the code to fix it for the future.

This is the function as it is minified:
```javascript
s.extend=function(a,b,c){if(b&&(b instanceof s||b.__dt_wrapper)){var e,d,f,h=function(a,b,c){return function(){var e=b.apply(a,arguments);s.extend(e,e,c.methodExt);return e}};e=0;for(d=c.length;e<d;e++)f=c[e],b[f.name]="function"===
typeof f.val?h(a,f.val,f):g.isPlainObject(f.val)?{}:f.val,b[f.name].__dt_wrapper=!0,s.extend(a,b[f.name],f.propExt)}};
```

This is the function unminified:
```javascript
_Api.extend = function ( scope, obj, ext )
	{
		// Only extend API instances and static properties of the API
		if ( ! obj || ( ! (obj instanceof _Api) && ! obj.__dt_wrapper ) ) {
			return;
		}
	
		var
			i, ien,
			j, jen,
			struct, inner,
			methodScoping = function ( scope, fn, struc ) {
				return function () {
					var ret = fn.apply( scope, arguments );
	
					// Method extension
					_Api.extend( ret, ret, struc.methodExt );
					return ret;
				};
			};
	
		for ( i=0, ien=ext.length ; i<ien ; i++ ) {
			struct = ext[i];
	
			// Value
			obj[ struct.name ] = typeof struct.val === 'function' ?
				methodScoping( scope, struct.val, struct ) :
				$.isPlainObject( struct.val ) ?
					{} :
					struct.val;
	
			obj[ struct.name ].__dt_wrapper = true;
	
			// Property extension
			_Api.extend( scope, obj[ struct.name ], struct.propExt );
		}
	};
```

A further point, you could add the following to the start of the _Api,extend function, as it appears to not do anything if ext's length is 0 (as there are hundreds of calls where length is 0 in my setup):
```javascript
if(!ext.length) {
return;
}
```

**Comment by kevdesign** [comment]
+1

**Comment by chrisdingli** [comment]
+1

**Comment by DataTables** [comment]
I've just been looking into this and struggling to reproduce the issue unless I have two Firefox windows open. I've put up two examples based on your code with the very latest DataTables code and without the JSFiddle shell code:

* min: http://datatables.net/dev/firefox-api/min.html
* debug: http://datatables.net/dev/firefox-api/debug.html

For both files I'm consistency getting around 600mS total execution time in the profile for each draw. Sorting the table multiple times shows the same result.

If I have multiple Firefox windows open (to test them both at the same time) then all bets are off - I have no idea what is going on there! The results for both fluctuate wildly.

Can anyone else tests these pages and say what you see?

**Comment by DataTables** [comment]
Good point about the early exit if `ext` is 0 length. I've [committed that](https://github.com/DataTables/DataTablesSrc/commit/60e3feb73d65).

**Comment by Benno007** [comment]
Thank you for looking into this :-)

Are you referring to separate windows or separate tabs in your testing? Kind of weird, when I have two windows open of the same URL, with the profiler docked to the window, the recording of the profile appears in BOTH, even though I'm only reloading one window... and it always seems to be roughly half the execution time :| O_O. There's no localStorage or indexDb or anything shared between windows?

I tried opening the min one in my Firefox just then and it ran at 2330.946ms 13236 calls with only one window open for the first time (in firebug).

My active plugins were Firebug and Web Developer toolbar, as well as FirePHP, open with and TamperData. I disabled all plugins and reloaded Firefox to the minified one, and I noticed something in the default inspector... Is it possible there's a memory leak?

As I continued sorting, the profiled time (in Firefox's default web inspector) continued climbing. When I came back to Chrome for a few minutes then went back to FF and sorted, it jumped from 700ms to 1300ms, and then after a few sorts it went back down again (probably a memory allocation thing for when the browser isn't in the foreground).

This is just with Firefox's default inspector and that minified version open.

![datatables](https://cloud.githubusercontent.com/assets/313195/6138494/537c2cf8-b1cd-11e4-9408-34bf9c4f4d7b.png)

But with Firebug, it seems to just show script execution time, and I can definitely see the difference between the profile on the minified one vs the debug one (in Firebug).

**Comment by DataTables** [comment]
Yes, two separate windows. I noticed the profiles showing up in both windows as well, and bizarrely they were slightly different (sounds like a bug to me :-) ).

I've deconstructed the min version of the API extend function and aside from semi-colon insertion I can't see that there would be any change between the min and debug version - there is no significant rewriting by Closure Compiler there.

It might be worth disabling Firebug and see if that helps? I've noticed before that it can seriously kill performance in Firefox.

**Comment by Benno007** [comment]
Hmm, I did try disabling Firebug and all other plugins, and it still seemed to be slow with the minified version.

Maybe firefox is just having trouble optimising the function internally with that dual ternary condition or something. Much weirds. Thank you for looking into this for me though.
